### PR TITLE
[2/n] Introduce node-protocol

### PR DESCRIPTION
**Pull request by AhmedSoliman**
[2/n] Introduce node-protocol

---
[//]: # (BEGIN SAPLING FOOTER)
Stack created with [Sapling](https://sapling-scm.com). Best reviewed with [ReviewStack](https://reviewstack.dev/restatedev/restate/pull/1242).
* #1245
* #1244
* #1243
* __->__ #1242

### CODE REVIEW

File: crates/node-protocol/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+publish = false
+
+[features]
+default = []
+
+[dependencies]
+restate-types = { workspace = true, features = ["serde"] }
+
+anyhow = { workspace = true, optional = true }
+bytes = { workspace = true, optional = true }
```
Comment (tillrohrmann) [inline crates/node-protocol/Cargo.toml:17]: Why adding these optional dependencies?

File: crates/node-protocol/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+derive_more = { workspace = true }
+enum-map = { workspace = true }
+prost = { workspace = true }
+prost-types = { workspace = true }
+serde = { workspace = true}
+strum = { workspace = true }
+strum_macros = { workspace = true }
+thiserror = { workspace = true, optional = true }
```
Comment (tillrohrmann) [inline crates/node-protocol/Cargo.toml:25]: Why adding `thiserror`? It seems not used in `node-protocol`.

**Review by tillrohrmann — APPROVED** [review]
Nice work @AhmedSoliman. I do understand now what you mean with directly reusing simple protobuf types. I think it makes sense how you used them to avoid adding more boilerplate. I also learned about `prost_build::Config::enum_attribute` which is really cool :-)

+1 for merging this PR.

**Comment by AhmedSoliman** [comment]
Manually merged

File: crates/node-protocol/src/metadata.rs [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+    const TARGET: TargetName = TargetName::MetadataManager;
+
+    fn kind(&self) -> &'static str {
+        self.into()
+    }
+}
+
+impl WireSerde for MetadataMessage {
+    fn encode(&self, _protocol_version: ProtocolVersion) -> Result<Bytes, CodecError> {
+        // serialize message to bytes
```
Comment (tillrohrmann) [inline crates/node-protocol/src/metadata.rs:46]: should we add an assert stating that `_protocol_version` is actually `BINCODE`?

**Review by tillrohrmann — APPROVED** [review]
Great work @AhmedSoliman. +1 for merging.

File: crates/core/Cargo.toml [hunk 4]
```diff
@@ -12,9 +12,11 @@ test-util = []
 
 [dependencies]
 restate-types = { workspace = true }
+restate-node-protocol = { workspace = true }
 
 anyhow = { workspace = true }
 arc-swap = { workspace = true }
+async-trait = { workspace = true }
```
[the remark this answers is not in the document]
Comment (AhmedSoliman) [reply crates/core/Cargo.toml:19]: Yes, that's used in a MessageRouter up in this stack of PRs.

**Pull request merged**